k8s-operator/sessionrecording/ws/message_test.go

```diff
@@ -0,0 +1,203 @@
+			b:               append([]byte{0x80, 0x83, 0x1, 0x2, 0x3, 0x4}, maskedBytes(testMask, []byte{0x1, 0x7, 0x8})...),
+			initialPayload:  []byte{0x1, 0x2, 0x3},
+			wantIsFinalized: true,
+			wantPayload:     []byte{0x1, 0x2, 0x3, 0x7, 0x8},
+			wantStreamID:    1,
+		},
+		{
+			name:            "single_large_fragment_no_mask_length_hint_126",
+			b:               append(append([]byte{0x80, 0x7e}, bs126Len...), append([]byte{0x1}, bs126...)...),
+			wantIsFinalized: true,
+			wantPayload:     bs126,
+			wantStreamID:    1,
+		},
+		{
+			name:            "single_large_fragment_no_mask_length_hint_127",
+			b:               append(append([]byte{0x80, 0x7f}, bs127Len...), append([]byte{0x1}, bs127...)...),
+			wantIsFinalized: true,
+			wantPayload:     bs127,
+			wantStreamID:    1,
+		},
+		{
+			name:    "zero_length_bytes",
+			b:       []byte{},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			msg := &message{
+				typ:     binaryMessage,
+				payload: tt.initialPayload,
+			}
+			if _, err := msg.Parse(tt.b, zl.Sugar()); (err != nil) != tt.wantErr {
+				t.Errorf("msg.Parse() = %v, wantsErr: %t", err, tt.wantErr)
+			}
+			if msg.isFinalized != tt.wantIsFinalized {
+				t.Errorf("wants message to be finalized: %t, got: %t", tt.wantIsFinalized, msg.isFinalized)
+			}
+			if msg.streamID.Load() != tt.wantStreamID {
+				t.Errorf("wants stream ID: %d, got: %d", tt.wantStreamID, msg.streamID.Load())
+			}
+			if !reflect.DeepEqual(msg.payload, tt.wantPayload) {
+				t.Errorf("unexpected message payload after Parse, wants %b got %b", tt.wantPayload, msg.payload)
+			}
+		})
+	}
+}
+
+// Test_msg_Parse_Rand calls Parse with a randomly generated input to verify
+// that it doesn't panic.
+func Test_msg_Parse_Rand(t *testing.T) {
+	zl, err := zap.NewDevelopment()
+	if err != nil {
+		t.Fatalf("error creating a test logger: %v", err)
+	}
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	for i := range 100 {
+		n := r.Intn(4096)
+		b := make([]byte, n)
+		_, err := r.Read(b)
+		if err != nil {
+			t.Fatalf("error generating random byte slice: %v", err)
+		}
+		msg := message{typ: binaryMessage}
+		f := func() {
+			msg.Parse(b, zl.Sugar())
+		}
+		testPanic(t, f, fmt.Sprintf("[%d] Parse panicked running with byte slice of length %d: %v", i, n, r))
+	}
+}
+
+func bytesSlice2ByteLen(t *testing.T) ([]byte, []byte) {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	var n uint16
+	n = uint16(rand.Intn(65535 - 1)) // space for and additional 1 byte stream ID
+	b := make([]byte, n)
+	_, err := r.Read(b)
+	if err != nil {
+		t.Fatalf("error generating random byte slice: %v ", err)
+	}
+	bb := make([]byte, 2)
+	binary.BigEndian.PutUint16(bb, n+1) // + stream ID
+	return b, bb
+}
+
+func byteSlice8ByteLen(t *testing.T) ([]byte, []byte) {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	var n uint64
+	n = uint64(rand.Intn(2 ^ 64 - 1 - 1)) // space for and additional 1 byte stream ID
```

Comment:
> ^ here is XOR, not exponent, so we're calling rand.Intn with 64 as the argument (0x2^0x40 == 0x42 == 66). However, if we fixed that we would be trying to allocate up to 2^64 bytes == 18 exabytes. We should probably limit it to a fixed constant barely big enough to trigger the 127 length indicator condition like 2^16 + 1

I have no recollection of what I was aiming for with that calculation

> Seems like we should error gracefully on much smaller payload sizes than we currently do to avoid that runtime panic when allocating a slice big enough to read the payload.

I was previously checking that the payload length is not more than `math.MaxInt64`, however looks like the actual max array length is smaller https://groups.google.com/g/golang-nuts/c/mw5f2kyEjpA

I found that x/net/websocket that serverside kube websocket implementation uses has a default max array size https://cs.opensource.google/go/x/net/+/refs/tags/v0.28.0:websocket/websocket.go;l=40, so now I'm checking against that (both in code when pre-creating an array for the payload, but also using the same in this test)
(From what I can see apimachinery does not modify the default max, so we can safely use that limit)
https://github.com/kubernetes/apimachinery/tree/master/pkg/util/httpstream/wsstream